ocs_ci/utility/utils.py

```diff
@@ -2514,13 +2516,80 @@ def check_ceph_health_not_ok():
     sampler.wait_for_func_status(True)
 
 
-def ceph_health_check(namespace=None, tries=20, delay=30):
+def ceph_health_resolve_daemon_crash():
+    """
+    Fix ceph health issue with daemon crash
+    """
+    log.warning("Trying to fix the issue with daemon crash by archiving crashes")
+    from ocs_ci.ocs.resources.pod import get_ceph_tools_pod
+
+    ct_pod = get_ceph_tools_pod()
+    ceph_crash_info_display(ct_pod)
+    archive_ceph_crashes(ct_pod)
+
+
+def ceph_health_recover(health_status, namespace=None):
+    """
+    Function which tries to recover ceph health to be HEALTH OK
+
+    Args:
+        health_status (str): Ceph health status
+        namespace (str): Namespace of OCS
+
+    Raises:
+        CephHealthExceptionNotRecovered: When Ceph health was not recovered
+        CephHealthExceptionRecovered: When Ceph health was recovered
+
+    """
+    ceph_health_fixes = [
+        {
+            "pattern": r"daemons have recently crashed",
+            "func": ceph_health_resolve_daemon_crash,
+            "func_args": [],
+            "func_kwargs": {},
+            "ceph_health_tries": 5,
+            "ceph_health_delay": 30,
+        },
+        # TODO: Add more patterns and fix functions
+    ]
+    for fix_dict in ceph_health_fixes:
+        pattern = fix_dict["pattern"]
+        if re.search(pattern, health_status):
+            log.info(
+                "Trying to fix Ceph Health becasue we found in Health status the matching pattern"
+                f": '{pattern}'!"
+            )
+            fix_dict["func"](
+                *fix_dict.get("func_args", []), **fix_dict.get("func_kwargs", {})
+            )
+            try:
+                ceph_health_check(
+                    namespace,
+                    tries=fix_dict.get("ceph_health_tries", 5),
+                    delay=fix_dict.get("ceph_health_delay", 30),
+                )
+            except Exception as ex:
+                log.error("")
+                raise CephHealthExceptionNotRecovered(
+                    f"Attempt to try to recover the Ceph Health failed! Exception: {ex}"
+                )
+            raise CephHealthExceptionRecovered(
```

Comment:
As we discussed I think we should handle this Exception if caught during the setup so that the test case can still execute. We can still raise this during the teardown of the test case so the information isn't lost but it also won't block the test case.

Reply:
I have updated the logic to not fail the test case at the setup - but rather store the exception and raise it in the teardown - so we will still have coverage and the test should have error in the teardown so it should be visible in RP and analyzed by owner of the test.